File: demos/47711/σ░ÅΦ▒åσ¡É/backend/app/schemas/commission.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator


CommissionType = Literal["seeker", "artist"]
ApplicationStatus = Literal["pending", "accepted", "rejected"]


class CommissionCreateRequest(BaseModel):
    type: CommissionType
    avatar: str | None = Field(default="", max_length=255)
    title: str = Field(min_length=1, max_length=120)
    desc: str = Field(min_length=1, max_length=1000)
    styles: list[str] = Field(default_factory=list, max_length=12)
    priceRange: str = Field(default="面议", max_length=100)
    turnaround: str = Field(default="待商议", max_length=100)
    samples: list[str] = Field(default_factory=list, max_length=12)

    @field_validator("avatar", "priceRange", "turnaround")
    @classmethod
    def normalize_text(cls, value: str | None) -> str:
        return (value or "").strip()

    @field_validator("title", "desc")
    @classmethod
    def validate_required_text(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("字段不能为空")
        return cleaned

    @field_validator("styles")
    @classmethod
    def normalize_styles(cls, value: list[str]) -> list[str]:
        styles: list[str] = []
        for item in value:
            cleaned = item.strip()
            if cleaned and cleaned not in styles:
                styles.append(cleaned[:20])
        return styles[:12]

    @field_validator("samples")
    @classmethod
    def normalize_samples(cls, value: list[str]) -> list[str]:
        samples: list[str] = []
        for item in value:
            cleaned = item.strip()
            if cleaned:
                samples.append(cleaned[:255])
        return samples[:12]


class CommissionApplyRequest(BaseModel):
    msg: str | None = Field(default=None, max_length=300)

    @field_validator("msg")
    @classmethod
    def normalize_msg(cls, value: str | None) -> str | None:
        if value is None:
            return None
        cleaned = value.strip()
        return cleaned or None


class CommissionApplicationOut(BaseModel):
    id: int
    user_id: int
    name: str
    msg: str | None = None
    time: datetime
    status: ApplicationStatus


class CommissionOut(BaseModel):
    id: int
    type: CommissionType
    avatar: str
    author: str
    title: str
    desc: str
    styles: list[str]
    priceRange: str
    turnaround: str
    samples: list[str]
    status: str
    time: datetime
    applicants: list[CommissionApplicationOut] = Field(default_factory=list)
    ownedByMe: bool = False
    appliedByMe: bool = False
    myApplicationId: int | None = None


class CommissionDashboardOut(BaseModel):
    received_applications: list[CommissionOut] = Field(default_factory=list)
    in_progress_as_client: list[CommissionOut] = Field(default_factory=list)
    in_progress_as_artist: list[CommissionOut] = Field(default_factory=list)
    pending_outgoing: list[CommissionOut] = Field(default_factory=list)

File: demos/47711/σ░ÅΦ▒åσ¡É/frontend/utils/apis/commission.js
import { request } from '../api.js'


function toTimestamp(value) {
	if (!value) return Date.now()
	if (typeof value === 'number') return value
	const parsed = new Date(value).getTime()
	return Number.isNaN(parsed) ? Date.now() : parsed
}

function unwrapListPayload(payload) {
	if (Array.isArray(payload)) return payload
	if (Array.isArray(payload?.data)) return payload.data
	return []
}

function unwrapObjectPayload(payload) {
	if (payload && typeof payload === 'object' && !Array.isArray(payload)) {
		if (payload.data && typeof payload.data === 'object' && !Array.isArray(payload.data)) {
			return payload.data
		}
		return payload
	}
	return {}
}

function normalizeApplication(raw = {}) {
	const userId = Number(raw.user_id ?? raw.userId ?? 0)
	return {
		id: Number(raw.id || 0),
		user_id: userId,
		userId,
		name: raw.name || '',
		msg: raw.msg || '',
		time: toTimestamp(raw.time),
		status: raw.status || 'pending',
	}
}

function normalizeCommission(raw = {}) {
	const applicants = Array.isArray(raw.applicants) ? raw.applicants.map(normalizeApplication) : []
	return {
		id: Number(raw.id || 0),
		type: raw.type || 'seeker',
		avatar: raw.avatar || '👤',
		author: raw.author || '',
		title: raw.title || '',
		desc: raw.desc || '',
		styles: Array.isArray(raw.styles) ? raw.styles : [],
		priceRange: raw.priceRange || raw.price_range || '面议',
		turnaround: raw.turnaround || '待商议',
		samples: Array.isArray(raw.samples) ? raw.samples : [],
		status: raw.status || 'open',
		time: toTimestamp(raw.time || raw.created_at),
		applicants,
		ownedByMe: !!raw.ownedByMe,
		appliedByMe: !!raw.appliedByMe,
		myApplicationId: raw.myApplicationId ?? null,
	}
}

function normalizeCommissionList(rawList = []) {
	return rawList.map(normalizeCommission)
}

export async function getCommissions() {
	const res = await request({ url: '/commissions', auth: true })
	return normalizeCommissionList(unwrapListPayload(res))
}

export async function createCommission(data) {
	const res = await request({
		url: '/commissions',
		method: 'POST',
		data,
		auth: true,
	})
	return normalizeCommission(unwrapObjectPayload(res))
}

export async function getCommissionDetail(id) {
	const res = await request({ url: `/commissions/${id}`, auth: true })
	return normalizeCommission(unwrapObjectPayload(res))
}

export async function deleteCommission(id) {
	await request({
		url: `/commissions/${id}`,
		method: 'DELETE',
		auth: true,
	})
}

export async function applyCommission(id, data = {}) {
	const res = await request({
		url: `/commissions/${id}/apply`,
		method: 'POST',
		data,
		auth: true,
	})
	return normalizeCommission(unwrapObjectPayload(res))
}

export async function acceptCommissionApplication(id, applicationId) {
	const res = await request({
		url: `/commissions/${id}/applications/${applicationId}/accept`,
		method: 'POST',
		auth: true,
	})
	return normalizeCommission(unwrapObjectPayload(res))
}

export async function rejectCommissionApplication(id, applicationId) {
	const res = await request({
		url: `/commissions/${id}/applications/${applicationId}/reject`,
		method: 'POST',
		auth: true,
	})
	return normalizeCommission(unwrapObjectPayload(res))
}

export async function getCommissionDashboard() {
	const res = await request({
		url: '/me/commission-dashboard',
		auth: true,
	})
	const data = unwrapObjectPayload(res)
	return {
		receivedApplications: normalizeCommissionList(data.received_applications || []),
		inProgressAsClient: normalizeCommissionList(data.in_progress_as_client || []),
		inProgressAsArtist: normalizeCommissionList(data.in_progress_as_artist || []),
		pendingOutgoing: normalizeCommissionList(data.pending_outgoing || []),
	}
}

